From ba0bd9e4c64688c9de7458c0c9c134d226af0321 Mon Sep 17 00:00:00 2001
From: Eric Sandeen <sandeen@redhat.com>
Date: Mon, 4 Aug 2014 19:03:28 -0400
Subject: [fs] btrfs: Add ctime/mtime update for btrfs device add/remove

Message-id: <1407179109-13714-322-git-send-email-sandeen@redhat.com>
Patchwork-id: 87282
O-Subject: [RHEL7.1 PATCH 321/422] btrfs: Add ctime/mtime update for btrfs device add/remove.
Bugzilla: 1071179
RH-Acked-by: Zach Brown <zab@redhat.com>

From: Qu Wenruo <quwenruo@cn.fujitsu.com>

commit 5a1972bd9fd4b2fb1bac8b7a0b636d633d8717e3
Author: Qu Wenruo <quwenruo@cn.fujitsu.com>
Date:   Wed Apr 16 17:02:32 2014 +0800

    btrfs: Add ctime/mtime update for btrfs device add/remove.

    Btrfs will send uevent to udev inform the device change,
    but ctime/mtime for the block device inode is not udpated, which cause
    libblkid used by btrfs-progs unable to detect device change and use old
    cache, causing 'btrfs dev scan; btrfs dev rmove; btrfs dev scan' give an
    error message.

    Reported-by: Tsutomu Itoh <t-itoh@jp.fujitsu.com>
    Cc: Karel Zak <kzak@redhat.com>
    Signed-off-by: Qu Wenruo <quwenruo@cn.fujitsu.com>
    Signed-off-by: Chris Mason <clm@fb.com>
Signed-off-by: Eric Sandeen <sandeen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/btrfs/volumes.c b/fs/btrfs/volumes.c
index 31ea6ad..dd07729 100644
--- a/fs/btrfs/volumes.c
+++ b/fs/btrfs/volumes.c
@@ -1452,6 +1452,22 @@ out:
  return ret;
 }
 
+/*
+ * Function to update ctime/mtime for a given device path.
+ * Mainly used for ctime/mtime based probe like libblkid.
+ */
+static void update_dev_time(char *path_name)
+{
+ struct file *filp;
+
+ filp = filp_open(path_name, O_RDWR, 0);
+ if (!filp)
+  return;
+ file_update_time(filp);
+ filp_close(filp, NULL);
+ return;
+}
+
 static int btrfs_rm_dev_item(struct btrfs_root *root,
         struct btrfs_device *device)
 {
@@ -1704,10 +1720,14 @@ int btrfs_rm_device(struct btrfs_root *root, char *device_path)
 
  ret = 0;
 
- /* Notify udev that device has changed */
- if (bdev)
+ if (bdev) {
+  /* Notify udev that device has changed */
   btrfs_kobject_uevent(bdev, KOBJ_CHANGE);
 
+  /* Update ctime/mtime for device path for libblkid */
+  update_dev_time(device_path);
+ }
+
 error_brelse:
  brelse(bh);
  if (bdev)
@@ -2146,6 +2166,8 @@ int btrfs_init_new_device(struct btrfs_root *root, char *device_path)
   ret = btrfs_commit_transaction(trans, root);
  }
 
+ /* Update ctime/mtime for libblkid */
+ update_dev_time(device_path);
  return ret;
 
 error_trans:
-- 
1.7.1